strawberryfields/tdm/tdmprogram.py

```diff
@@ -0,0 +1,232 @@
+            spatial_modes += len(modes)
+
+    if not spatial_modes:
+        raise ValueError("Must be at least one measurement.")
+
+    if spatial_modes is not len(N):
+        raise ValueError("Number of measurements operators must match number of spatial_modes.")
+
+    return spatial_modes
+
+
+def input_check(args, copies):
+    """Checks the input arguments have consistent dimensions"""
+    # check if all lists are of equal length
+    param_lengths = [len(param) for param in args]
+    if len(set(param_lengths)) != 1:
+        raise ValueError("Gate-parameter lists must be of equal length.")
+
+    # check copies
+    if not isinstance(copies, int) or copies < 1:
+        raise TypeError("Number of copies must be a positive integer.")
+
+
+def _get_mode_order(num_of_values, N):
+    """Get the order by which the modes where measured"""
+    N = N.copy()
+
+    all_modes = []
+    for i in range(len(N)):
+        ra = list(range(sum(N[:i]), sum(N[: i + 1])))
+        all_modes.append(ra * ceil(max(N) / len(ra)))
+
+    mode_order = [i for j in zip(*all_modes) for i in j]
+    mode_order *= ceil(num_of_values / len(mode_order))
+
+    return mode_order[:num_of_values]
+
+
+def reshape_samples(all_samples, modes, N):
+    """Reshapes the samples dict so that they have the expected correct shape
+
+    Reshapes the samples dict so that they the measured modes are the ones
+    defined to be measured in the circuit instead of being spread over a larger
+    number of modes due to the mode-shifting occuring in TDMProgram.
+
+    Args:
+        all_samples (dict[int, list]): the raw measured samples
+        modes (Sequence[int]): the modes that are measured in the circuit
+        N (Sequence[int]): the number of used modes per belt/spatial modes
+
+    Returns:
+        new_samples (dict[int, list]): the re-shaped samples
+    """
+    num_of_values = len([i for j in all_samples.values() for i in j])
+    mode_order = _get_mode_order(num_of_values, N)
+
+    # go backwards through all_samples and add them into the correct mode
+    new_samples = dict()
+    for i, m in enumerate(mode_order):
+        idx = modes[i % len(N)]
+        if idx not in new_samples:
+            new_samples[idx] = []
+        new_samples[idx].append(all_samples[m].pop(0))
+    return new_samples
+
+
+class TDMProgram(sf.Program):
+    """Represents a photonic quantum circuit in the time domain encoding"""
+
+    def __init__(self, N, name=None):
+        # N is a list with the number of qumodes for each spatial mode
+        if isinstance(N, int):
+            self.N = [N]
+        else:
+            self.N = N
+        self.concurr_modes = sum(self.N)
+
+        self.total_timebins = 0
+        self.spatial_modes = 0
+        self.measured_modes = []
+        super().__init__(num_subsystems=self.concurr_modes, name=name)
+
+    # pylint: disable=arguments-differ, invalid-overridden-method
+    def context(self, *args, copies=1, shift="end"):
```

Comment:
The string `"end"` only made sense as long as we had `shift="after"` as an alternative. How about `"auto"` or `"standard"` or similar to distinguish from a user-defined shift (which takes place at the end as well)?

Reply:
`default`?